File: android/app/src/main/kotlin/pt/ualg/miaugenda/data/model/Shift.kt
package pt.ualg.miaugenda.data.model

data class ShiftUser(
    val id: Int,
    val name: String,
    val employeeNumber: String,
    val role: String,
    val category: String = ""
)

data class ShiftType(
    val id: Int,
    val name: String,
    val startTime: String,
    val endTime: String
)

data class Shift(
    val id: Int,
    val userId: Int,
    val shiftTypeId: Int? = null,
    val date: String,
    val startTime: String? = null,
    val endTime: String? = null,
    val published: Boolean = false,
    val user: ShiftUser? = null,
    val shiftType: ShiftType? = null
)

fun Shift.resolvedStartTime(): String = startTime ?: shiftType?.startTime ?: "00:00"
fun Shift.resolvedEndTime(): String   = endTime   ?: shiftType?.endTime   ?: "00:00"
fun Shift.resolvedName(): String      = shiftType?.name ?: "${resolvedStartTime()}-${resolvedEndTime()}"

File: desktop/src/models/shift.ts
export interface ShiftUser {
  id: number;
  name: string;
  employeeNumber: string;
  role: string;
  category: string;
}

export interface ShiftType {
  id: number;
  name: string;
  startTime: string;
  endTime: string;
}

export interface Shift {
  id: number;
  userId: number;
  shiftTypeId: number | null;
  date: string;
  startTime: string | null;
  endTime: string | null;
  published: boolean;
  user: ShiftUser | null;
  shiftType: ShiftType | null;
}

export interface CreateShiftRequest {
  userId: number;
  date: string;
  startTime: string;
  endTime: string;
  shiftTypeId?: number | null;
}

export interface UpdateShiftRequest {
  userId?: number;
  date?: string;
  startTime?: string;
  endTime?: string;
  shiftTypeId?: number | null;
  published?: boolean;
}

export function resolvedStartTime(shift: Shift): string {
  return shift.startTime ?? shift.shiftType?.startTime ?? "00:00";
}

export function resolvedEndTime(shift: Shift): string {
  return shift.endTime ?? shift.shiftType?.endTime ?? "00:00";
}

export function resolvedName(shift: Shift): string {
  return shift.shiftType?.name ?? `${resolvedStartTime(shift)}-${resolvedEndTime(shift)}`;
}